### ISSUE TITLE
Setting to control serialisation of None fields

### ISSUE THREAD

**Issue by robeady** [wyfo/apischema#178]
Let's say I want to avoid serialising dataclass fields that have `None` values. To make round-tripping work we can assume `None` is the default value for the field as well.

From reading the documentation there seem to be two options:
1. Decorate the class with `@with_fields_set`, and do not set the value at construction time
2. Annotate the field `: Union[T, UndefinedType] = Undefined` instead of `: Optional[T] = None`

Unfortunately both of these approaches require changes to the dataclass. 2 is also particularly awkward for code that uses the class and the behaviour of 1 is rather subtle. Is there a way to configure the serialiser to obtain this behaviour without touching the class? And in such a way that generating a serialisation JSON schema would understand that the field cannot be `null`?

**Comment by wyfo** [comment]
I agree that both options are not very convenient for your use case. 

By the way, there is a third option which is documented here : [Optional vs. NotNull](https://wyfo.github.io/apischema/data_model/#optional-vs-notnull); it requires PEP 593, i.e. either Python 3.8+ or `typing_extensions` installed.

However, you have to know that `apischema.skip.NotNull` will be kind of deprecated (but still available) in version 0.16.
In fact, if `NotNull` is ok for deserialization (I originally used it for configuration parsing), there is an issue when it comes to serialization: the field will be serialized as `null` (or raise an error if [serialized types are checked](https://wyfo.github.io/apischema/de_serialization/#type-checking)).

With the next version of *apischema* and conditional serialization, there will be a metadata (currently named [`apischema.metadata.none_as_undefined`](https://github.com/wyfo/apischema/blob/master/examples/none_as_undefined.py)) which will surely suit your needs better: it makes *apischema* consider the field as not nullable for deserialization, and skip serialization if its value is `None`. 

I'm quite late for for the release of 0.16, i'm working on a huge refactoring of performance optimizations, but I hope I will release before the end of august.

I know that handling of non-nullable optional fields is kind of messy currently, it's not an easy problem, and I'm open to alternative solutions; I hope `none_as_undefined` will be enough convenient to solve it.

**Comment by robeady** [comment]
Would there be a way to employ the none_as_undefined feature globally, rather than adding it to every relevant class field? Some of these dataclasses are in other libraries in my case and I would rather avoid modifying them

**Comment by wyfo** [comment]
> Would there be a way to employ the none_as_undefined feature globally, rather than adding it to every relevant class field? Some of these dataclasses are in other libraries in my case and I would rather avoid modifying them

I understand the issue of modifying fields one by one, especially for external libraries.

#180 should solve your issue. I've let the PR open if you wanna take a look at it (feedbacks are welcome). I will then merge it and it will be included in 0.16.

> edit: I am wondering if replacing the default_object_fields function would enable this although I am not sure exactly what the code there would have to do...

It was not possible before #181 to override dataclass fields with `default_object_fields`, but now it can be done (in 0.16) and it's indeed a possible solution. It would looks like:
```python
from dataclasses import dataclass, replace
from typing import Any, Optional, Sequence

from apischema import serialize, settings
from apischema.metadata import skip
from apischema.objects import ObjectField, object_fields

prev_default_object_fields = settings.default_object_fields


def default_object_fields(tp: Any) -> Sequence[ObjectField]:
    try:
        return [
            replace(f, metadata=skip(serialization_if=lambda v: v is None) | f.metadata)
            # setting default parameter is very important to prevent recursion error,
            # as object_fields would call settings.default_object_fields otherwise
            for f in object_fields(tp, default=prev_default_object_fields).values()
        ]
    except TypeError:  # tp is not an object type
        return prev_default_object_fields(tp)


settings.default_object_fields = default_object_fields


@dataclass
class A:
    a: Optional[str] = None


assert serialize(A, A()) == {}
```

Of course, with #181, `settings.serialization.exclude_none=True` is a lot easier.
